apollo-compiler/src/test/graphql/com/example/inline_fragment_inside_inline_fragment/TestQuery.kt

```diff
@@ -122,64 +122,139 @@ class TestQuery : Query<TestQuery.Data, Operation.Variables> {
         val name: String
 
         override fun marshaller(): ResponseFieldMarshaller
-      }
 
-      interface Human : Search {
-        override val __typename: String
+        interface Human : Character {
+          override val __typename: String
 
-        /**
-         * The home planet of the human, or null if unknown
-         */
-        val homePlanet: String?
+          /**
+           * The name of the character
+           */
+          override val name: String
 
-        override fun marshaller(): ResponseFieldMarshaller
+          /**
+           * The home planet of the human, or null if unknown
+           */
+          val homePlanet: String?
+
+          override fun marshaller(): ResponseFieldMarshaller
+        }
+
+        interface Droid : Character {
```

Comment:
I would drop the `Character`inheritance ( and `__typename` and `name` fields here) to just keep:

```
interface Droid {
  val primaryFunction: String?
}
```

That would simplify the codegen a bunch. There's a minor annoyance that users won't be able to do `hero.asDroid()?.name` anymore but they can still achieve the same results with something like

```
when (hero) {
  is DroidHero -> // do something with hero.name and hero.primaryFunction
}
```

It might actually simplify the mental model since inline fragments would be closer to named fragments

Reply:
We can't do that as we merge parent fields from Character, so any fields that defined in parent field / fragment will be merged down to Droid interface. Otherwise we can easily run into clash of fields when build implementation for multiple fragments, like for instance:
```
interface Droid {
  friends: List<Friend>

  interface Friend {
    ...
  }
}

interface Character {
 friends: List<Friend>

  interface Friend {
    ...
  }
}

data class DroidCharacterHero(val friends: List<Friend>): Character, Droid

```

here we going to clash on friends as it ambiguous type

Reply:
> here we going to clash on friends as it ambiguous type

Since we're going to generate a data class for `Friend`, we could implement the different interfaces there?

```
interface Droid {
  friends: List<Friend>

  interface Friend {
    ...
  }
}

interface Character {
 friends: List<Friend>

  interface Friend {
    ...
  }
}

data class DroidCharacter(val friends: List<Friend>): Character, Droid {
  data class Friend([...]) : Character.Friend, Droid.Friend
}
```

This is more symmetrical.

Reply:
hm, you are right.... let me try this again, I don't remember what was the issue and why this didn't work out last time, smth related to nested fragments. I think we can merge this PR and I will try to optimize this in next one.

Reply:
Sounds good to me!